Use hostname not netloc

ddormer opened the pull request:
Fixes #4275 and fixes #4158 by passing only the hostname/IP to urllib's `proxy_bypass` function which was timing out further down in `socket.gethostbyname` when passed a hostname or IP that contained authentication credentials.

While fixing the parameter passed to `proxy_bypass`, I also took the chance to replace other uses of `netloc` with `hostname` in the `should_bypass_function`.

nateprewitt commented on requests/utils.py:
```diff
@@ -679,28 +679,28 @@ def should_bypass_proxies(url, no_proxy):
                     # matches the IP of the index
                     return True
         else:
+            host_with_port = '{}:{}'.format(parsed.hostname, parsed.port)
```
A couple notes here. The `port` attribute is either an `int` or `None`, which means a case like http://localhost being passed to `urlparse` would yield `localhost:None` here. While that's unlikely to affect matching, I'm not sure that's what we want to be doing.

The second thing is we're still planning to support python 2.6 until 3.0.0 is released which means this format statement won't work as is. We'll need to provide something along the lines of `{0}:{1}` instead to avoid the ValueError.

ddormer commented:
@nateprewitt Thanks for the review! Below I've listed the changes I made before I squashed my commits.

- Check if the port exists before adding it to prevent `host:None`.
- Use `{0}` to support Python 2.6.
- Updated the HISTORY.rst.

sigmavirus24 commented:
@ddormer Can you rebase this so it's up-to-date with master?

ddormer commented:
@sigmavirus24 Thanks for reviewing this, I've rebased so it's up to date with master.
